### PR TITLE
[SPARK-23568][ML] Use metadata numAttributes if available in Silhouette

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

Silhouette need to know the number of features. This was taken using `first` and checking the size of the vector. Despite this works fine, if the number of attributes is present in metadata, we can avoid to trigger a job for this and use the metadata value. This can help improving performances of course.

## How was this patch tested?

existing UTs + added UT

### CODE REVIEW

**Comment by mgaido91** [comment]
@holdenk @sethah @srowen @viirya may you please help reviewing this PR if you have time? Thanks.

**Review by srowen — APPROVED** [review]
I'm not so familiar with this, but it looks logical

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 1]
```diff
@@ -170,6 +171,13 @@ private[evaluation] abstract class Silhouette {
   def overallScore(df: DataFrame, scoreColumn: Column): Double = {
     df.select(avg(scoreColumn)).collect()(0).getDouble(0)
   }
+
+  protected def getNumberOfFeatures(dataFrame: DataFrame, columnName: String): Int = {
+    val group = AttributeGroup.fromStructField(dataFrame.schema(columnName))
+    group.numAttributes.getOrElse {
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: We should use `size` of `AttributeGroup` to determine its number of attributes. When it is -1, we can leverage `first` then.

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**